File: leet_easy_c++/leet_235/leet_235a.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == NULL) return NULL;
        TreeNode* lefty = lowestCommonAncestor(root->left, p, q);
        TreeNode* righty = lowestCommonAncestor(root->right, p, q);
        if((root == p || root == q) && righty) return root;
        if((root == p || root == q) && lefty) return root;
        if(righty && lefty) return root;
        if(root == p || root == q) return root;
        if(lefty == NULL)
            return righty;
        else
            return lefty;
    }
};

File: leet_easy_javascript/leet_235/leet_235a.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

var which_node;

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
    which_node = null;
    lCAh(root, p, q);
    return which_node;
};

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
//~ lCAh = lowest Common Ancestor helper
var lCAh = function(root, p, q) {
    if(root === null) return null;
    var _left = null, _right = null;
    if(root === p){
        _left = lCAh(root.left, p, q);
        _right = lCAh(root.right, p, q);
        if( _left === q || _right === q) which_node = root;
        return root;
    }else if(root === q){
        _left = lCAh(root.left, p, q);
        _right = lCAh(root.right, p, q);
        if( _left === p || _right === p) which_node = root;
        return root;
    }else{
        _left = lCAh(root.left, p, q);
        _right = lCAh(root.right, p, q);
        if(_left !== null && _right !== null){
            which_node = root;
            return root;
        } 
        return _left || _right;
    }  
};